### PR TITLE
Add SpatialClassNLLCriterion

### PR DESCRIPTION
**Pull request by apaszke**
A CPU counterpart for torch/cunn#250

### CODE REVIEW

**Comment by apaszke** [comment]
One thing I'm not 100% sure about is `omp` pragma in `updateGradInput`. Are both loop counters private in the way I wrote it?

File: lib/THNN/generic/SpatialClassNLLCriterion.c [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  real *gradInput_data = THTensor_(data)(gradInput);
+
+  long batch_size = THTensor_(size)(input, 0);
+  long n_classes = THTensor_(size)(input, 1);
+  long map_size = THTensor_(size)(input, 2) * THTensor_(size)(input, 3);
+  long sample_size = map_size * n_classes;
+
+  real normalize = sizeAverage ? *total_weight_data : 1.0f;
+
+#pragma omp parallel for
```
Comment (soumith) [inline lib/THNN/generic/SpatialClassNLLCriterion.c]: this should be: omp parallel for private(b)

Comment (fmassa) [inline lib/THNN/generic/SpatialClassNLLCriterion.c]: I don't know much about omp and stuff, but I thought that the `private (b)` was only needed if `b` was defined outside of the scope of the for loop (a la C89). But I might be completely wrong.

Comment (apaszke) [inline lib/THNN/generic/SpatialClassNLLCriterion.c]: Yeah, that's what I understood as well (I.e. every variable declared in parallel block is private to the thread). Initially I had private(b, elem) to be sure, but GCC crashed on undefined references there if I didn't move the declarations above the loops.

Comment (soumith) [inline lib/THNN/generic/SpatialClassNLLCriterion.c]: you guys are right. sorry, my bad.

**Pull request merged**